### ISSUE TITLE
When encryption is activated, aria2c can't decode the torrent file

### ISSUE THREAD

**Issue by Depado** [e-alfred/ocdownloader#14]
Hi !
I've been wondering why aria2c kept throwing me those :

```
08/28 11:59:02 [DEBUG] Exception caught
Exception: [download_helper.cc:291] errorCode=25 Bencode decoding failed
```
Turns out, ocdownloader gives the filepath to aria2c, aria2c receives an encrypted torrent file and can't do much about this... 

Is there a quick way I can fix this, without disabling encryption in Nextcloud ?

**Comment by Nibbels** [comment]
Nice catch!

**Comment by Nibbels** [comment]
My next question is whether or not your files get encrypted when downloaded.

**Comment by Depado** [comment]
I don't understand your question 😞  
I tried uploading a torrent file directly form ocdownloader's interface, looks like the issue is the same, as soon as the file is received it is encrypted, then the path is sent to aria which can't make sense of the data it is given.

**Comment by Nibbels** [comment]
Well when we just write a file to your downloads folder, then nextcloud has to read and encrypt and save it somehow. Else it would be not encrypted. This might not be a problem at all but we need to consider this.

I remember reading some documentation .. and found it again:
https://docs.nextcloud.com/server/12/admin_manual/configuration_files/encryption_configuration.html
https://docs.nextcloud.com/server/12/admin_manual/_images/encryption3.png
"Files are encrypted while uploading"

There are commands for encrypting **some users only** but I dont see one to encrypt / have decrypted **some folders only**.
if we or you might find some _occ command for that_ I might be able to find the function for decrypting one file/folder and adapt it for those .torrents

**Comment by Depado** [comment]
Yes that's one limitation of the encryption module, it's all or nothing.
According to [the documentation](https://docs.nextcloud.com/server/12/admin_manual/configuration_files/encryption_configuration.html#files-not-encrypted) "Third-party app data" aren't encrypted. Maybe there is a way to store the uploaded torrent files as app data ? I don't know much about third party apps (and just started using nextcloud a few weeks ago).

**Comment by Depado** [comment]
[This may be useful](https://docs.nextcloud.com/server/12/developer_manual/app/appdata.html)

**Comment by Nibbels** [comment]
aria2 would be able to have a link sent instead of the torrent, i guess...

**Comment by Depado** [comment]
That would indeed be a solution :) Would it be possible to add a field for pasting a torrent link in ocdownloader ?

**Comment by Depado** [comment]
In fact the "better" option would be to add :
- One field for torrent links
- One field for magnet links

I'll have a look at how ocdownloader is coded, I don't know much in PHP so not sure if I'm going to be any help with that, but I'll try automatically check if the encryption module is enabled and if so, how to write torrents to an app data folder. (Or even to a temporary unencrypted folder)

**Comment by Nibbels** [comment]
I am totally with you!
My time got a bit limited these days. But that is a good solution I think.
For me less fields within the UI look better:
_If local file then use local file
if magnet then use magnet
if .torrent in link then download..._

**Comment by Nibbels** [comment]
there is some function:
/httpdocs/apps/encryption/lib/Recovery.php
https://github.com/nextcloud/server/blob/master/apps/encryption/lib/Recovery.php#L304
or this for folders?
https://github.com/nextcloud/server/blob/master/apps/encryption/lib/Recovery.php#L282

private function recoverFile($path, $privateKey, $uid) {

**Comment by e-alfred** [comment]
I think there could be an additional Bittorrent download screen allowing people to enter both Magnet links and links to *.torrent files. Depending on how the link starts (magnet: or http/https:), Aria2 would then be started in the right mode.

**Comment by Depado** [comment]
That would be awesome 😃

**Comment by TheWanderingCoel** [comment]
Any Progress on this? I'm facing the same issue with server side encryption enabled.
```
2020-01-13 15:08:25.721496 [INFO] [HttpListenCommand.cc:78] RPC: Accepted the connection from 127.0.0.1:47480.
2020-01-13 15:08:25.721594 [INFO] [HttpServer.cc:187] HTTP Server received request
POST /jsonrpc HTTP/1.1
Host: 127.0.0.1:6800
Accept: */*
Content-Length: 63209
Content-Type: application/x-www-form-urlencoded
Expect: 100-continue


2020-01-13 15:08:26.722989 [INFO] [SaveSessionCommand.cc:74] No change since last serialization or startup. No serialization is necessary this time.
2020-01-13 15:08:26.723390 [INFO] [rpc_helper.cc:103] Executing RPC method aria2.addTorrent
2020-01-13 15:08:26.725601 [INFO] [RpcMethodImpl.cc:302] Uploaded torrent data was saved as /home/nextcloud/xxxxxx/files/Downloads/xxx.mkv/1x5cxxxfxx3x00xbxxxxxxxxxxe712bf.torrent
2020-01-13 15:08:26.725942 [DEBUG] [RpcMethod.cc:101] 捕捉到异常
Exception: [download_helper.cc:322] errorCode=25 Bencode decoding failed
2020-01-13 15:08:26.726043 [DEBUG] [HttpServer.cc:296] HTTP Server sends response:
HTTP/1.1 400 Bad Request
Date: Mon, 13 Jan 2020 07:08:26 GMT
Content-Length: 92
Expires: Mon, 13 Jan 2020 07:08:26 GMT
Cache-Control: no-cache
Content-Type: application/json-rpc
Access-Control-Allow-Origin: *
Connection: close


2020-01-13 15:08:26.726161 [INFO] [AbstractHttpServerResponseCommand.cc:118] CUID#12 - HttpServer: all response transmitted.
```

**Comment by anguschang007** [comment]
Any possible solution for this issue? I am facing the same issue, there will come out an error message "**Returned GID is null ! Is Aria2c running as a daemon ?**" when I am trying to use the encrypted torrent file to start the download task. However, it is fine if the torrent is not encrypted. 
Also, I also tried to use the "New Magnet/HTTP download" to download the torrent file, and think that it will start to download the torrent's content automatically. But ocdownloader doesn't start the download's progress properly, the Download folder will appear the torrent's content with the correct filesize but ocdownloader will not have the task for torrent and I use the iotop to check the hard drive IO but no any read/write there. It means the torrent's content in the Download folder is fake content.
